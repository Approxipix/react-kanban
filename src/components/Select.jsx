import React, { Component } from 'react';
import { Label } from './BaseComponent';
import styled from 'styled-components'

const Wrapper = styled.div`
  position: relative;
  margin-bottom: .5rem;
`;

const Dropdown = styled.div`
  position: absolute;
  top: calc(100% + .5rem);
  left: 0;
  width: 100%;
  max-height: 12rem;
  box-shadow: 0 .25rem .3rem 0 rgba(4, 19, 23, 0.22);
  background-color: #ebecf0;
  opacity: ${props => props.isOpened ? '1' : '0'};
  visibility: ${props => props.isOpened ? 'visible' : 'hidden'};
  transform: translateY(-.5rem);
  transition: .2s all ease-in-out;
  will-change: transform;
  z-index: 20;
  
`;

const DropdownItem = styled.button`
  width: 100%;
  padding: 0 .75rem;
  outline: none;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  background-color: transparent;
  text-align: left;
  color: #172b4d;
  transition: all .2s ease-in;
  &:hover {
    color: #fff;
    background-color: #2E7EAF;
  }
`;

const Input = styled.input`
  position: relative;
  min-height: 2rem;
  width: 100%;
  padding: 0 .75rem;
  border: 0;
  background-color: #ebecf0;
  box-shadow: 0 1px 0 0 rgba(9, 30, 66, .13);
  color: #172b4d;
  font-size: 1rem;
  line-height: 1.5rem;
  border-radius: .25rem;
  cursor: pointer;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: #dfe1e6;
  }
`;

const DropdownTitle = styled.p``;

class Select extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
    };
  }

  toggleDropdown = (isOpened) => {
    this.setState({
      isOpened: typeof isOpened !== 'undefined' ? isOpened : !this.state.isOpened,
    });
  };

  render() {
    let { isOpened } = this.state;
    let {
      value,
      label,
      options,
      onChange,
      placeholder,
    }  = this.props;
    return (
      <Wrapper onBlur={() => this.toggleDropdown(false)}>
        <Label>
          {label}
          <Input
            type="text"
            value={value || ''}
            readOnly={true}
            placeholder={placeholder}
            onClick={() => this.toggleDropdown()}
            onChange={() => this.toggleDropdown(true)}
          />
        </Label>
        <Dropdown isOpened={isOpened}>
          {options.map((option, i) => (
            <DropdownItem
              key={`${option.value}-${i}`}
              type="button"
              tabIndex={-1}
              onMouseDown={() => {
                onChange(option.value);
                this.toggleDropdown(false);
              }}
            >
              <DropdownTitle>
                {option.title}
              </DropdownTitle>
            </DropdownItem>
          ))}
        </Dropdown>
      </Wrapper>
    );
  }
}
export default Select;
