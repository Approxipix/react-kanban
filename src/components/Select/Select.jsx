import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Label } from '../BaseComponent';
import styled from 'styled-components'

const Wrapper = styled.div`
  position: relative;
  margin-bottom: .5rem;
`;

const Dropdown = styled.div`
  position: absolute;
  top: calc(100% + .5rem);
  left: 0;
  width: 100%;
  max-height: 12rem;
  box-shadow: 0 .25rem .3rem 0 rgba(4, 19, 23, 0.22);
  background-color: #ebecf0;
  opacity: ${props => props.isOpened ? '1' : '0'};
  visibility: ${props => props.isOpened ? 'visible' : 'hidden'};
  transform: translateY(-.5rem);
  transition: .2s all ease-in-out;
  will-change: transform;
  z-index: 20;
`;
Dropdown.displayName = 'Dropdown';

const DropdownItem = styled.button`
  width: 100%;
  padding: 0 .75rem;
  outline: none;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  background-color: transparent;
  text-align: left;
  color: #172b4d;
  transition: all .2s ease-in;
  &:hover {
    color: #fff;
    background-color: #2E7EAF;
  }
`;

const Input = styled.input`
  position: relative;
  min-height: 2rem;
  width: 100%;
  padding: 0 .75rem;
  border: 0;
  background-color: #ebecf0;
  box-shadow: 0 1px 0 0 rgba(9, 30, 66, .13);
  color: #172b4d;
  font-size: 1rem;
  line-height: 1.5rem;
  border-radius: .25rem;
  cursor: pointer;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: #dfe1e6;
  }
`;
Input.displayName = 'Input';

const DropdownTitle = styled.p``;

class Select extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
    };
  }

  toggleDropdown = (isOpened) => {
    this.setState({
      isOpened: typeof isOpened !== 'undefined'
        ? isOpened
        : !this.state.isOpened,
    });
  };

  render() {
    let {
      value,
      label,
      options,
      onChange,
      placeholder,
    }  = this.props;
    let { isOpened } = this.state;
    return (
      <Wrapper onBlur={() => this.toggleDropdown(false)}>
        <Label>
          {label}
          <Input
            type="text"
            name={label}
            value={value}
            readOnly={true}
            placeholder={placeholder}
            onClick={() => this.toggleDropdown()}
            onChange={() => this.toggleDropdown(true)}
          />
        </Label>
        <Dropdown isOpened={isOpened}>
          {options && options.map((option, i) => (
            <DropdownItem
              key={`${option.value}-${i}`}
              type="button"
              tabIndex={-1}
              onMouseDown={() => {
                onChange(option.value);
                this.toggleDropdown(false);
              }}
            >
              <DropdownTitle>
                {option.title}
              </DropdownTitle>
            </DropdownItem>
          ))}
        </Dropdown>
      </Wrapper>
    );
  }
}

Select.defaultProps = {
  value: '',
  label: '',
  options: [],
  onChange: () => {},
  placeholder: '',
};

Select.propTypes = {
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      value: PropTypes.oneOfType(
        [PropTypes.string, PropTypes.number]
      ).isRequired,
    })
  ).isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired,
};

export default Select;
