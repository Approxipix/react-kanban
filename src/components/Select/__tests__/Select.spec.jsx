import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import Select from '../Select';

describe('<Select>', () => {
  let component;
  let mockOnChange;

  beforeEach(() => {
    mockOnChange = jest.fn();
    const props = {
      value: 'Test value',
      label: 'Test label',
      options: [
        { title: 'Option 1', value: 'Option1Value' },
        { title: 'Option 2', value: 'Option2Value' },
        { title: 'Option 3', value: 3 },
      ],
      onChange: mockOnChange,
      placeholder: 'Test placeholder',
    };
    component = shallow(<Select {...props} />);
  });

  it('should renders correctly without props', () => {
    const component = shallow(<Select />);

    expect(toJson(component)).toMatchSnapshot()
  });

  it('should renders without crashing given the required props', () => {
    expect(toJson(component)).toMatchSnapshot()
  });

  it('should render a dropdown item for each option', () => {
    expect(component.find('Dropdown').children().length).toEqual(3)
  });

  it("should call toggleDropdown on input change and click", () => {
    const spy = jest.spyOn(component.instance(), "toggleDropdown");
    component.instance().forceUpdate();
    const mockEvent = {
      target: {
        name: "Test label",
        value: "Test"
      }
    };

    component.find('Input').simulate("change", mockEvent);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(true);

    component.find('Input').simulate("click");
    expect(spy).toHaveBeenCalledTimes(2);
  });

  it('checks if toggleDropdown sets isOpened correctly', () => {
    component.instance().toggleDropdown();
    expect(component.instance().state.isOpened).toEqual(true);

    component.instance().toggleDropdown(false);
    expect(component.instance().state.isOpened).toEqual(false);
  });

  it('should sets the isOpened prop on the Dropdown', () => {
    expect(component.find('Dropdown').props().isOpened).toEqual(false);
    component.instance().toggleDropdown();
    expect(component.find('Dropdown').props().isOpened).toEqual(true)
  });

  it("should call onChange after click on dropdown item", () => {
    component.find('Dropdown').childAt(0).simulate('mousedown');
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith('Option1Value');
  });
});
