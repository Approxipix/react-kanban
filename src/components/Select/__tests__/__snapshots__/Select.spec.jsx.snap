// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select> should renders correctly without props 1`] = `
<styled.div
  onBlur={[Function]}
>
  <styled.label>
    <Input
      name=""
      onChange={[Function]}
      onClick={[Function]}
      placeholder=""
      readOnly={true}
      type="text"
      value=""
    />
  </styled.label>
  <Dropdown
    isOpened={false}
  />
</styled.div>
`;

exports[`<Select> should renders without crashing given the required props 1`] = `
<styled.div
  onBlur={[Function]}
>
  <styled.label>
    Test label
    <Input
      name="Test label"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="Test placeholder"
      readOnly={true}
      type="text"
      value="Test value"
    />
  </styled.label>
  <Dropdown
    isOpened={false}
  >
    <styled.button
      key="Option1Value-0"
      onMouseDown={[Function]}
      tabIndex={-1}
      type="button"
    >
      <styled.p>
        Option 1
      </styled.p>
    </styled.button>
    <styled.button
      key="Option2Value-1"
      onMouseDown={[Function]}
      tabIndex={-1}
      type="button"
    >
      <styled.p>
        Option 2
      </styled.p>
    </styled.button>
    <styled.button
      key="3-2"
      onMouseDown={[Function]}
      tabIndex={-1}
      type="button"
    >
      <styled.p>
        Option 3
      </styled.p>
    </styled.button>
  </Dropdown>
</styled.div>
`;
